pyogrio/tests/test_geopandas_io.py

```diff
@@ -187,40 +190,105 @@ def test_read_fids_force_2d(test_fgdb_vsi):
 
 
 @pytest.mark.parametrize(
-    "driver,ext",
-    [
-        ("ESRI Shapefile", "shp"),
-        ("GeoJSON", "geojson"),
-        ("GeoJSONSeq", "geojsons"),
-        ("GPKG", "gpkg"),
-    ],
+        "suffix",
+        [".fgb", ".geojson", ".geojsons", ".gpkg", ".json", ".shp", ],
 )
-def test_write_dataframe(tmpdir, naturalearth_lowres, driver, ext):
-    expected = read_dataframe(naturalearth_lowres)
-
-    filename = os.path.join(str(tmpdir), f"test.{ext}")
-    write_dataframe(expected, filename, driver=driver)
-
-    assert os.path.exists(filename)
-
-    df = read_dataframe(filename)
-
-    if driver != "GeoJSONSeq":
-        # GeoJSONSeq driver I/O reorders features and / or vertices, and does
-        # not support roundtrip comparison
+def test_write_dataframe(tmp_path, naturalearth_lowres, suffix):
+    input_gdf = read_dataframe(naturalearth_lowres)
+    assert isinstance(input_gdf, gp.GeoDataFrame)
+    output_path = tmp_path / f"test{suffix}"
+
+    if suffix == ".fgb":
+        # For FlatGeoBuf mixed types are not supported + input_gdf is mixed
+        with pytest.raises(Exception, match="Could not add feature to layer at"):
+            write_dataframe(input_gdf, output_path, promote_to_multitype=False)
+        write_dataframe(input_gdf, output_path)
+    else:
+        write_dataframe(input_gdf, output_path)
+
+    assert output_path.exists()
+    result_gdf = read_dataframe(output_path)
+    assert isinstance(result_gdf, gp.GeoDataFrame)
+
+    # File types not supporting mixed types should only contain 1 type
+    geometry_types = result_gdf.geometry.type.unique()
+    if suffix in [".fgb", ".gpkg", ]:
+        assert len(geometry_types) == 1
+    else:
+        assert len(geometry_types) == 2
+
+    if suffix == ".geojsons":
```

Comment:
`_vectorized.normalize` => `pg.normalize`.

This does indeed add a bit of complexity to get around sort / multi related issues.

I agree on your points:
> the sort_geodataframe function would be available as a parameter in assert_geodataframe_is_equal
the to_multipolygon function would be available in GeoPandas (or rather a version that supports all geometry types)

Temporarily adding them here seems OK.  @jorisvandenbossche might have opinions about how best to proceed there.

Reply:
Indeed, for the testing there is the existing `check_like` keyword to ignore the order of rows/columns, but that does assume you have row and column labels to sort by. And that is not the case here for the row labels, I suppose, as it is just a default range (and rows in different order).

Agreed that something like this would be useful to have in `assert_geodataframe_equal`, but for now having a helper here should be fine.

Reply:
In this specific case, though, the naturalearth_lowres has a column "name" with unique values, so sorting on just that column is sufficient, which would simplify the code quite a bit.

Reply:
True, I removed the redundant code.